lib/java-qmstr/src/main/java/org/qmstr/util/FilenodeUtils.java

```diff
@@ -88,61 +88,28 @@ public static String getHash(JarFile jarfile, JarEntry jarEntry) {
             String[] filename = relSrcPath.getFileName().toString().split("\\.");
             filename[filename.length-1] = "class";
             String className = String.join(".", filename);
-            Path packageDirs = relSrcPath.getParent();
+            Path packageDirs = relSrcPath.getParent() != null ? relSrcPath.getParent() : Paths.get(".");
             Path classesRelPath = packageDirs.resolve(className);
 
-            if (packageDirs != null) {
-                return outDirs.stream()
-                        .filter(od -> FilenodeUtils.isActualClassDir(od, classesRelPath))
-                        .map(outdir -> {
-                            Path classesPath = outdir.toPath().resolve(classesRelPath);
-                            Set<Path> nested = FilenodeUtils.getNestedClasses(outdir.toPath().resolve(packageDirs), filename[filename.length - 2]);
-                            nested.add(classesPath);
-                            return nested.stream()
-                                    .map(p -> FilenodeUtils.getFileNode(p, FilenodeUtils.getTypeByFile(p.getFileName().toString())))
-                                    .map(node -> node.toBuilder().addDerivedFrom(sourceNode).build())
-                                    .collect(Collectors.toSet());
-                        }).flatMap(sets -> sets.stream())
-                        .collect(Collectors.toSet());
-            }
+            return outDirs.stream()
+                    .filter(od -> FilenodeUtils.isActualClassDir(od, classesRelPath))
+                    .map(outdir -> {
+                        Path classesPath = outdir.toPath().resolve(classesRelPath);
+                        Set<Path> nested = FilenodeUtils.getNestedClasses(outdir.toPath().resolve(packageDirs), filename[filename.length - 2]);
+                        nested.add(classesPath);
+                        return nested.stream()
+                                .map(p -> FilenodeUtils.getFileNode(p, FilenodeUtils.getTypeByFile(p.getFileName().toString())))
+                                .map(node -> node.toBuilder().addDerivedFrom(sourceNode).build())
+                                .collect(Collectors.toSet());
+                    }).flatMap(sets -> sets.stream())
+                    .collect(Collectors.toSet());
+
```

Comment:
I just want to comment on this function in general. Instead of passing "Collection<File> sourceDirs" and getting the first and then throwing FNF exception, how about changing it to "File sourceDir" and require the caller to pass one only valid sourceDir?

Otherwise let's bubble up the exception.

Reply:
Bubble up the exception is not possible right now as this is called in stream processing. But we should find a better solution for this, later.